### PR TITLE
[C-API] Modify element-wise control functions @open sesame 7/28 15:15

### PR DESCRIPTION
**Pull request by again4you**
To comply with Tizen API Design Guideline, this PR uses single key-value pair
for `ml_pipeline_element_set_property()` and `ml_pipeline_element_get_property()` functions.

It contains the following patches.
* [C-API] Modify element-wise control functions
* [Test/CAPI] Update testcases of element-wise control functions

Signed-off-by: Sangjung Woo <sangjung.woo@samsung.com>

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed []Skipped
2. Run test: [*]Passed [ ]Failed []Skipped

**Related ACR**
* https://code.sec.samsung.net/jira/browse/ACR-1571

#### Submit v2
* Make a separate getter and setter for each type to comply with Tizen API Design Guideline.
* Use fixed width integer types (i.e. `int32_t`, `int64_t`) instead of primitive type (i.e. `int`, `long`).
* Need more scenario test cases.

#### Submit v1
* First draft

### CODE REVIEW

File: api/capi/include/nnstreamer.h [hunk 1]
```diff
@@ -600,156 +606,131 @@ int ml_pipeline_valve_set_open (ml_pipeline_valve_h valve_handle, bool open);
...
+  }
+
+  error:
+    ml_pipeline_element_release_handle (elem_h);
+    ml_pipeline_destroy (handle);
+    g_free(pipeline);
  * @endcode
  */
-int ml_pipeline_element_set_property (ml_pipeline_element_h elm_h, const char *first_property_name, ...);
+int ml_pipeline_element_set_property (ml_pipeline_element_h elem_h, const char *property_name, ml_value_type value);
```
Comment (myungjoo) [inline api/capi/include/nnstreamer.h:680]: Can ```value``` be const?

File: api/capi/src/nnstreamer-capi-pipeline.c [hunk 2]
```diff
@@ -1646,103 +1646,84 @@ ml_pipeline_element_get_handle (ml_pipeline_h pipe, const char *element_name,
...
+    ml_loge ("The second argument, property name is not valid.");
+    ret = ML_ERROR_INVALID_PARAMETER;
+    goto unlock_return;
+  }
 
-  /* Check input parameter */
-  if (first_property_name == NULL) {
-    ml_loge ("The second argument, first property name is not valid.");
+  if (value == NULL) {
+    ml_loge ("The third argument, value is not valid.");
```
Comment (myungjoo) [inline api/capi/src/nnstreamer-capi-pipeline.c:1676]: Uh... can't you do ```set_property (elem, "whatever-int-property", (void *) 0);``` ?

File: api/capi/src/nnstreamer-capi-pipeline.c [hunk 3]
```diff
@@ -1646,103 +1646,84 @@ ml_pipeline_element_get_handle (ml_pipeline_h pipe, const char *element_name,
...
+    ml_loge ("The property name [%s] does not exist.", property_name);
+    ret = ML_ERROR_INVALID_PARAMETER;
+    goto unlock_return;
   }
-  va_end (args_check);
 
-  g_object_set_valist (G_OBJECT (elem->element), first_property_name, args);
-  va_end (args);
+  /* Set property */
+  g_object_set (G_OBJECT (elem->element), property_name, value, NULL);
```
Comment (myungjoo) [inline api/capi/src/nnstreamer-capi-pipeline.c:1689]: ```g_object_set_data()```?

File: api/capi/src/nnstreamer-capi-pipeline.c [hunk 4]
```diff
@@ -1646,103 +1646,84 @@ ml_pipeline_element_get_handle (ml_pipeline_h pipe, const char *element_name,
...
+    ml_loge ("The property name [%s] does not exist.", property_name);
+    ret = ML_ERROR_INVALID_PARAMETER;
+    goto unlock_return;
   }
-  va_end (args_check);
 
   /* Get property */
-  g_object_get_valist (G_OBJECT (elem->element), first_property_name, args);
-  va_end (args);
+  g_object_get (G_OBJECT (elem->element), property_name, value, NULL);
```
Comment (myungjoo) [inline api/capi/src/nnstreamer-capi-pipeline.c:1724]: ```s/value/&value``` or do ```value = g_object_get_data(...)```.

Comment (myungjoo) [reply api/capi/src/nnstreamer-capi-pipeline.c:1724]: And you probably need more test cases (it should've failed)

File: api/capi/include/nnstreamer.h [hunk 5]
```diff
@@ -124,6 +124,12 @@ typedef void *ml_pipeline_element_h;
  */
 typedef void *ml_custom_easy_filter_h;
 
+/**
+ * @brief A generic type that can hold values of any other type
```
Comment (jaeyun-jung) [inline api/capi/include/nnstreamer.h:128]: missing dot

File: api/capi/include/nnstreamer.h [hunk 6]
```diff
@@ -600,156 +606,131 @@ int ml_pipeline_valve_set_open (ml_pipeline_valve_h valve_handle, bool open);
...
- * ml_pipeline_element_release_handle (elem_h);
- * ml_pipeline_destroy (handle);
- * g_free(pipeline);
+ *  ml_pipeline_element_release_handle (elem_h);
+ *  ml_pipeline_destroy (handle);
+ *  g_free(pipeline);
  * @endcode
  */
-int ml_pipeline_element_get_property (ml_pipeline_element_h elm_h, const char *first_property_name, ...);
+int ml_pipeline_element_get_property (ml_pipeline_element_h elem_h, const char *property_name, ml_value_type value);
```
Comment (jaeyun-jung) [inline api/capi/include/nnstreamer.h:733]: Q: is it possible to set void * type ?
int ml_pipeline_element_get_property (ml_pipeline_element_h elem_h, const char *property_name, void *value);
int ml_pipeline_element_set_property (ml_pipeline_element_h elem_h, const char *property_name, const void *value);

File: api/capi/src/nnstreamer-capi-pipeline.c (hunk 2 above)
Comment (again4you) [reply api/capi/src/nnstreamer-capi-pipeline.c:1676]: Right!

File: api/capi/include/nnstreamer.h (hunk 5 above)
Comment (again4you) [reply api/capi/include/nnstreamer.h:128]: ACK.

File: api/capi/include/nnstreamer.h (hunk 6 above)
Comment (again4you) [reply api/capi/include/nnstreamer.h:733]: Yes, it's possible. In this case, developers should explicitly cast a primitive value as below.

```cpp
// API Header
int ml_pipeline_element_set_property (ml_pipeline_element_h elem_h, const char *property_name, const void *value);

// Developer's code
status = ml_pipeline_element_set_property(valve_h, "drop", (const void*)TRUE);
status = ml_pipeline_element_set_property(selector_h, "sync-mode", (const void*)1);
```

`g_object_set_data()` also has the same type-casting problem so it uses `GINT_TO_POINTER` macro as below.
```cpp
// Link: https://github.com/GNOME/glib/blob/mainline/gio/gio-tool-mount.c#L198
g_object_set_data (G_OBJECT (op), "state", GINT_TO_POINTER (MOUNT_OP_ABORTED));
```

File: tests/tizen_capi/unittest_tizen_capi.cc [hunk 7]
```diff
@@ -5289,661 +5289,51 @@ TEST (nnstreamer_capi_singleshot, invoke_dynamic_fail_n)
...
-  EXPECT_EQ (status, ML_ERROR_NONE);
-
-  status = ml_pipeline_element_release_handle (selector_h);
-  EXPECT_EQ (status, ML_ERROR_NONE);
-
-  status = ml_pipeline_destroy (handle);
-  EXPECT_EQ (status, ML_ERROR_NONE);
+  error:
+    ml_pipeline_element_release_handle (elem_h);
+    ml_pipeline_destroy (handle);
```
Comment (dongju-chae) [inline tests/tizen_capi/unittest_tizen_capi.cc:5335]: How about initializing `elem_h` and `handle` to `NULL`?

For example, I think the goto statement in line 5307 makes some problem due to uninitialized `elem_h`

File: api/capi/src/nnstreamer-capi-pipeline.c [hunk 8]
```diff
@@ -1671,103 +1671,271 @@ ml_pipeline_element_get_handle (ml_pipeline_h pipe, const char *element_name,
...
+    ml_loge ("The property name [%s] does not exist.", property_name);
+    return FALSE;
+  }
+
+  /* Compare property's type with given type */
+  if (!((pspec->value_type == type) ||
+    (type == G_TYPE_ENUM && G_TYPE_IS_ENUM (pspec->value_type)) ||
+    (type == G_TYPE_INT64 && pspec->value_type == G_TYPE_LONG) ||
+    (type == G_TYPE_UINT64 && pspec->value_type == G_TYPE_ULONG) ||
+    (type == G_TYPE_DOUBLE && pspec->value_type == G_TYPE_FLOAT))) {
```
Comment (jaeyun-jung) [inline api/capi/src/nnstreamer-capi-pipeline.c:1704]: value-type is G_TYPE_DOUBLE or G_TYPE_FLOAT?

Comment (again4you) [reply api/capi/src/nnstreamer-capi-pipeline.c:1704]: It's OK since I alreay check `pspec->value_type == type` first.

**Comment by again4you** [comment]
`Review process` is done so it's time to merge.
* https://code.sec.samsung.net/jira/browse/ACR-1571

**Comment by myungjoo** [comment]
I think it's ok, but let the last issue raiser, @jaeyun-jung confirm it.

**Pull request merged**